File: 0207_CourseSchedule/CourseSchedule.py
import queue
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        inD = [0]*numCourses
        adjList = [list() for _ in range(numCourses)]
        for cur, pre in prerequisites:
            adjList[pre].append(cur)
            inD[cur]+=1
        S = 0
        q = queue.Queue()
        for i in range(numCourses):
            if inD[i] == 0:
                q.put(i)

        while not q.empty():
            tmp = q.get()
            S+=1
            for c in adjList[tmp]:
                inD[c]-=1
                if not inD[c]: 
                    q.put(c)
        return S == numCourses

File: 0207_CourseSchedule/CourseSchedule.cxx
#include<vector>
#include<queue>
using namespace std;

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> inD(numCourses, 0);
        vector< vector<int> > adjList( numCourses, vector<int>() );
        for(auto && p : prerequisites){
            adjList[p[1]].push_back(p[0]);
            inD[p[0]]++;
        }
        int sum = 0;
        queue<int> q;
        for(int i=0; i<numCourses; ++i){
            if(!inD[i]) q.push(i);
        }

        while(!q.empty()){
            auto c = q.front();
            q.pop();
            sum++;
            for(auto & it: adjList[c]){ 
                inD[it]--;
                if(!inD[it]) q.push(it);
            }
        }
        return sum == numCourses;

    }
};

